rmf_task_ros2/src/rmf_task_ros2/Dispatcher.cpp

```diff
@@ -0,0 +1,414 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include <rmf_task_ros2/Dispatcher.hpp>
+
+#include <rclcpp/node.hpp>
+
+#include "action/Client.hpp"
+
+#include <rmf_task_msgs/srv/submit_task.hpp>
+#include <rmf_task_msgs/srv/cancel_task.hpp>
+#include <rmf_task_msgs/srv/get_task_list.hpp>
+
+namespace rmf_task_ros2 {
+
+//==============================================================================
+class Dispatcher::Implementation
+{
+public:
+  std::shared_ptr<rclcpp::Node> node;
+  std::shared_ptr<bidding::Auctioneer> auctioneer;
+  std::shared_ptr<action::Client> action_client;
+
+  using SubmitTaskSrv = rmf_task_msgs::srv::SubmitTask;
+  using CancelTaskSrv = rmf_task_msgs::srv::CancelTask;
+  using GetTaskListSrv = rmf_task_msgs::srv::GetTaskList;
+
+  rclcpp::Service<SubmitTaskSrv>::SharedPtr submit_task_srv;
+  rclcpp::Service<CancelTaskSrv>::SharedPtr cancel_task_srv;
+  rclcpp::Service<GetTaskListSrv>::SharedPtr get_task_list_srv;
+
+  StatusCallback on_change_fn;
+
+  DispatchTasks active_dispatch_tasks;
+  DispatchTasks terminal_dispatch_tasks;
+  std::size_t i = 0; // temp index for generating task_id
```

Comment:
One more suggestion that I overlooked before: We should give this a more explicit name than just `i`. It's far too easy for a for-loop to accidentally shadow this member variable.

I would suggest a name like `task_counter`.